src/OrbitBase/JoinFutures.cpp

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitBase/JoinFutures.h"
+
+#include <memory>
+
+#include "OrbitBase/Future.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Promise.h"
+#include "absl/synchronization/mutex.h"
+
+namespace orbit_base {
+namespace {
+struct SharedStateJoin {
+  Promise<void> promise;
+  absl::Mutex mutex;
+  int uncompleted_futures;
+};
+}  // namespace
+
+Future<void> JoinFutures(absl::Span<const Future<void>> futures) {
+  auto shared_state = std::make_shared<SharedStateJoin>();
+  absl::MutexLock lock{&shared_state->mutex};
+
+  for (const auto& future : futures) {
+    CHECK(future.IsValid());
+    orbit_base::FutureRegisterContinuationResult const result =
+        future.RegisterContinuation([shared_state]() {
+          absl::MutexLock lock{&shared_state->mutex};
+          --shared_state->uncompleted_futures;
+
+          if (shared_state->uncompleted_futures == 0) {
+            shared_state->promise.MarkFinished();
+          }
+        });
+
+    if (result == orbit_base::FutureRegisterContinuationResult::kSuccessfullyRegistered) {
+      ++shared_state->uncompleted_futures;
+    }
+  }
+
+  if (shared_state->uncompleted_futures == 0) {
```

Comment:
maybe check features.empty() at line 24 and do early return?

Reply:
I debated this with myself and decided against it. But now I don't see why anymore. Maybe the code was different at that point.

Fixed. Thanks!